franklin/github/views.py

```diff
@@ -136,42 +65,52 @@ def repository_list(request):
+        # TODO - #53 will have us refactoring all errors returned from our
+        # external dependencies. It will need to handle cases like this where
+        # we get a potential non-error like this one.
         serializer = SiteSerializer(data=request.data)
         if serializer and serializer.is_valid():
+            try:
+                site = Site.objects.get(github_id=request.data['github_id'])
+                return Response('Resource already exists', status=422)
+            except Site.DoesNotExist:
+                pass  # Expected
             site = serializer.save()
             if not request.user.details.has_repo_access(site):
                 message = 'Current user does not have permission for this repo'
                 logger.warn(message + ' | %s | %s', request.user, site)
                 return Response(message, status=status.HTTP_403_FORBIDDEN)
-        config = get_franklin_config(site, request.user)
-        if config and not hasattr(config, 'status_code'):
-            # Optional. Update DB with any relevant .franklin config
-            pass
-
-        # TODO - The below 422 checks are temporary fixes. #78 will have this
-        # happen less as we will delete webhooks from github when we remove
-        # repos from franklin-api. #53 will have us refactoring all errors
-        # returned from our external dependencies. It will need to handle cases
-        # like this where we get a potential non-error like this one.
-        webhook_response = create_repo_webhook(site, request.user)
-        if not status.is_success(webhook_response.status_code):
-            if webhook_response.status_code != 422:
-                # This is some error other than webhook already exists
-                return Response(status=webhook_response.status_code)
-        deploy_key_response = create_repo_deploy_key(site, request.user)
-        if not status.is_success(deploy_key_response.status_code):
-            if deploy_key_response.status_code != 422:
-                # This is some error other than deploy_key already exists
-                return Response(status=deploy_key_response.status_code)
-
-        return Response(status=status.HTTP_201_CREATED)
+
+        retrieve_franklin_config_file(site, request.user)
+
+        # Call Github to register a webhook
+        webhook_r = create_repo_webhook(site, request.user)
+        if (status.is_success(webhook_r.status_code) or
+                webhook_r.status_code == 422):
+            site.webhook_id = webhook_r.json().get('id', '')
+            # Call Github to register a deploy key
+            deploy_key_r = create_repo_deploy_key(site, request.user)
+            if (status.is_success(deploy_key_r.status_code) or
+                    deploy_key_r.status_code == 422):
+                site.deploy_key_id = deploy_key_r.json().get('id', '')
+                site.save()
+                return Response(status=status.HTTP_201_CREATED)
+        delete_site(site, request.user)
```

Comment:
If we had any issues calling github during registration, we will delete everything and return an error to the user. This logic is a bit too complicated and will be refactored soon.